compiler/rustc_lint_defs/src/builtin.rs

```diff
@@ -3377,9 +3377,36 @@ declare_lint_pass! {
         IMPLIED_BOUNDS_ENTAILMENT,
         BYTE_SLICE_IN_PACKED_STRUCT_WITH_DERIVE,
         AMBIGUOUS_GLOB_REEXPORTS,
+        LONG_RUNNING_CONST_EVAL,
     ]
 }
 
+declare_lint! {
+    /// The `long_running_const_eval` lint is emitted when const
+    /// eval is running for a long time to ensure rustc terminates
+    /// even if you accidentally wrote an infinite loop.
+    ///
+    /// ### Example
+    ///
+    /// ```rust,compile_fail
+    /// const FOO: () = loop {};
+    /// ```
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// Loops allow const evaluation to compute arbitrary code, but may also
+    /// cause infinite loops or just very long running computations.
+    /// Without this lint, there is no output from the compiler indicating that
+    /// anything is happening.
+    /// Users can opt into long running computations by allowing the lint
```

Comment:
Given how there is an unsilenceable warning now a little bit after the lint fires, this comment should maybe be updated to not reccomend this to users any more?

Reply:
I don't understand. We should stop recommending to allow the lint? You need to allow the lint in order to get a long running const eval (and the warnings). If the lint is `deny`, const eval just stops before the first warning is emitted.

Reply:
Allowing the lint is neccessary, but it's not sufficient to get arbitrarily complex constant evaluation without warnings. There is no way to disable those warnings. Therefore I think it's creating false expectations if you advertise disabling the lint as opting in: when I have opted into something I don't want to be warned about it in addition.

The ability to disable the lint is more for temporary workarounds or something.

I'm not suggesting any functional changes here, the way this PR is implemented is good. I'm just saying that maybe this part of the docs should be adjusted as it might confuse readers about the intent (unless I misunderstood the reasoning behind why this PR is implemented in the way it is).

A proposal:

> Users can enable long running computations by allowing the lint on individual constants or for entire crates.
> #### Unconditional warnings
> Note that regardless of whether the lint is allowed or set to warn, the compiler will issue warnings if constant evaluation runs significantly longer than this lint's limit. These warnings are also shown to downstream users from crates.io or similar registries. If you are above the lint's limit, both you and downstream users might be exposed to these warnings. They might also appear on compiler updates, as the compiler makes minor changes about how complexity is measured: staying below the limit ensures that there is enough room, and given that the lint is disabled for people who use your dependency it means you will be the first to get the warning and can put out an update before it affects users.

Maybe it can be shortened a little but in general I think this information is very useful for users as it changes the meaning of the prior sentence a lot.

Reply:
I took most of your text and modified it a bit